src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -0,0 +1,276 @@
+            @Override public long writeHtmlTo(long start, Writer w) throws IOException {
+                try (BufferedReader indexBR = index.isFile() ? Files.newBufferedReader(index.toPath(), StandardCharsets.UTF_8) : new BufferedReader(new NullReader(0))) {
+                    ConsoleAnnotationOutputStream<FlowExecutionOwner.Executable> caos = new ConsoleAnnotationOutputStream<>(w, ConsoleAnnotators.createAnnotator(build), build, StandardCharsets.UTF_8);
+                    long r = this.writeRawLogTo(start, new FilterOutputStream(caos) {
+                        long lastTransition = -1;
+                        boolean eof;
+                        String lastId;
+                        long pos = start;
+                        boolean hadLastId;
+                        @Override public void write(int b) throws IOException {
+                            while (lastTransition < pos && !eof) {
+                                String line = indexBR.readLine();
+                                if (line == null) {
+                                    eof = true;
+                                    break;
+                                }
+                                int space = line.indexOf(' ');
+                                try {
+                                    lastTransition = Long.parseLong(space == -1 ? line : line.substring(0, space));
+                                } catch (NumberFormatException x) {
+                                    LOGGER.warning("Ignoring corrupt index file " + index);
+                                }
+                                lastId = space == -1 ? null : line.substring(space + 1);
+                            }
+                            if (pos == lastTransition) {
+                                if (hadLastId) {
+                                    w.write(LogStorage.endStep());
+                                }
+                                hadLastId = lastId != null;
+                                if (lastId != null) {
+                                    w.write(LogStorage.startStep(lastId));
+                                }
+                            }
+                            super.write(b);
+                            pos++;
+                        }
+                        @Override public void flush() throws IOException {
+                            if (lastId != null) {
+                                w.write(LogStorage.endStep());
+                            }
+                            super.flush();
+                        }
+                    });
+                    ConsoleAnnotators.setAnnotator(caos.getConsoleAnnotator());
+                    return r;
+                }
+            }
+        };
+    }
+
+    @Override public AnnotatedLargeText<FlowNode> stepLog(FlowNode node, boolean complete) {
+        String id = node.getId();
```

Comment:
@jglick Similarly, while this has a few comments it is not clear what the intent is without reverse-engineering the intent from what it's doing.

Reply:
+1. Yeah, meaningful and granular comments is going to help code maintainability.